xcode/Subconscious/Shared/Services/AddressBookService.swift

```diff
@@ -75,27 +88,85 @@ class AddressBookService {
             .eraseToAnyPublisher()
     }
     
-    func followUser(did: Did, petname: Petname) throws {
+    func hasEntryForPetname(petname: Petname) throws -> Bool {
+        return try getPetname(petname: petname) != nil
+    }
+    
+    func hasEntryForPetnameAsync(petname: Petname) -> AnyPublisher<Bool, Error> {
+        CombineUtilities.async(qos: .default) {
+            return try self.hasEntryForPetname(petname: petname)
+        }
+    }
+    
+    func isFollowingUser(did: Did, petname: Petname) throws -> Bool {
+        guard let user = try getPetname(petname: petname) else {
+            return false
+        }
+        
+        return user == did
+    }
+    
+    func isFollowingUserAsync(did: Did, petname: Petname) -> AnyPublisher<Bool, Error> {
+        CombineUtilities.async(qos: .default) {
+            return try self.isFollowingUser(did: did, petname: petname)
+        }
+    }
+    
+    func findAvailablePetname(petname: Petname) throws -> Petname {
+        var name = petname
+        var count = 0
+        let MAX = 99
+        
+        while try hasEntryForPetname(petname: name) {
+            guard let next = petname.increment() else {
+                throw AddressBookError.exhaustedUniquePetnameRange
+            }
+           
+            name = next
+            count += 1
+            
+            // Escape hatch, no infinite loops plz
+            if count > MAX {
+                throw AddressBookError.exhaustedUniquePetnameRange
+            }
+        }
+        
+        return name
+    }
+    
+    func findAvailablePetnameAsync(petname: Petname) -> AnyPublisher<Petname, Error> {
+        CombineUtilities.async(qos: .default) {
+            return try self.findAvailablePetname(petname: petname)
+        }
+    }
+    
+    func followUser(did: Did, petname: Petname, preventOverwrite: Bool = false) throws {
         if try self.noosphere.identity() == did.id {
             throw AddressBookError.cannotFollowYourself
         }
         
+        if try hasEntryForPetname(petname: petname) && preventOverwrite {
+            throw AddressBookError.invalidAttemptToOverwitePetname
+        }
```

Comment:
It seems like you should be able to put `preventOverwrite` first here to shortcircuit the logic, but Swift scolds me and says `try` cannot be next to `&&` (even with bracketing).

I chose this && format over a nested if statement, but it has _slightly_ worse performance characteristics like this.